### PR TITLE
sb-04: - done;

### PR DESCRIPTION
**Pull request by jah701**
* controllers implemented (comment, product, user);
* all necessary functionality to find data implemented;
* tests implemented;

### CODE REVIEW

File: src/main/java/com/spring/boot/controller/CommentController.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+@RequestMapping("/comments")
+public class CommentController {
+    private final CommentService commentService;
+
+    @Autowired
+    public CommentController(CommentService commentService) {
+        this.commentService = commentService;
+    }
+
+    @GetMapping("/most_used_words/{amount}")
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/spring/boot/controller/CommentController.java]: https://restfulapi.net/resource-naming/

File: src/main/java/com/spring/boot/controller/ProductController.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+@RequestMapping("/products")
+public class ProductController {
+    private final ProductService productService;
+
+    @Autowired
+    public ProductController(ProductService productService) {
+        this.productService = productService;
+    }
+
+    @GetMapping("/most_commented/{amount}")
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/spring/boot/controller/ProductController.java]: same

File: src/main/java/com/spring/boot/controller/UserController.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+@RequestMapping("/users")
+public class UserController {
+    private final UserService userService;
+
+    @Autowired
+    public UserController(UserService userService) {
+        this.userService = userService;
+    }
+
+    @GetMapping("/most_active/{amount}")
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/spring/boot/controller/UserController.java]: and here

File: src/main/java/com/spring/boot/repository/ProductRepository.java [hunk 4]
```diff
@@ -1,9 +1,16 @@
...
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface ProductRepository extends JpaRepository<Product, Long> {
+    @Query(value = "SELECT c.product_id FROM comments c "
+            + "GROUP BY c.product_id "
+            + "ORDER BY count(c.product_id) DESC "
+            + "LIMIT ?1", nativeQuery = true)
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/spring/boot/repository/ProductRepository.java]: optional: try to use JPQL

File: src/main/java/com/spring/boot/repository/UserRepository.java [hunk 5]
```diff
@@ -1,9 +1,17 @@
...
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface UserRepository extends JpaRepository<User, Long> {
+    @Query(value = "SELECT * FROM Users u "
+            + "WHERE u.helpfulness_Numerator > '0' "
+            + "ORDER BY u.helpfulness_Numerator DESC "
+            + "LIMIT ?1", nativeQuery = true)
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/spring/boot/repository/UserRepository.java]: try to use JPA possibilities Spring will give you hints and probably you will have a method with a really long name but without SQL at all.

File: src/main/java/com/spring/boot/repository/CommentRepository.java [hunk 6]
```diff
@@ -1,9 +1,19 @@
...
+import java.util.Optional;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface CommentRepository extends JpaRepository<Comment, Long> {
+    @Query(value = "SELECT c FROM Comment c "
+            + "WHERE c.productId = ?1")
+    List<Optional<Comment>> getProductComments(String productId);
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/spring/boot/repository/CommentRepository.java]: I don't see the point of fetching Optionals. (you don't check them anyway) Se what you will get if there is no result (empty list or exception)

File: src/main/java/com/spring/boot/service/impl/CommentServiceImpl.java [hunk 7]
```diff
@@ -22,7 +28,48 @@ public void add(Comment comment) {
...
+        String[] splitWords = wordsFromRepository
+                .toString()
+                .toLowerCase()
+                .split("\\W+");
+
+        Map<String, List<String>> words = getGroupedMap(splitWords);
+
+        Map<String, List<String>> collect = getSortedMap(words);
+
+        return getLimitedResultList(collect, amount);
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/spring/boot/service/impl/CommentServiceImpl.java]: let's do this with the power of SQL or event JPQL.

File: src/test/java/util/CommentServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    static void beforeAll() {
+        commentRepository = Mockito.mock(CommentRepository.class);
+        commentService = new CommentServiceImpl(commentRepository);
+    }
+
+    @Test
+    public void outputListSize_ok() {
+        List<String> tempList = List.of("the ", "and ", "this ", "a ",
+                "it ", "i ", "taffy ", "is ", "was ", "very ", "of ");
+        when(commentService.getMostUsedWordsFromComments(11)).thenReturn(tempList);
```
Comment (dmytro-hryhoruk) [inline src/test/java/util/CommentServiceImplTest.java]: if you are testing the service method, there is no point in mocking this service. Mock repository

**Review by dmytro-hryhoruk — APPROVED** [review]

File: src/test/java/util/CommentServiceImplTest.java (hunk 8 above)
Comment (jah701) [reply src/test/java/util/CommentServiceImplTest.java]: I do mock repo not service

**Pull request merged**